soc: xtensa: add support for specifying the service port

enjiamai opened the pull request:
This PR includes adding parameters for supporting:

- Specifying the service port for west / twister /hardware map
  (This also change the behavior that west flash have to add --pty {host} to show the output immediately.) 
- Specifying the service port for the cavs server
- Add --tool-opt for the west to pass to cavstool 
- Update the zephyr doc for these changes.
- Remove the dependency of netifaces python package.


**Details:**
Add specifying the port for the remote cavstool remote service.
Ex.
  
```
       west flash --remote-host {host}:{port} \
                  --pty {host}:{port}
```

Specify the port is optional when running west.

And another major change is now we have to specify the remote log
service by --pty to make it output the log message immediately.
Previously it will output directly. Why we make this change is
because this is more close to the behavior we use west flash.
```
west flash --remote-host=192.168.0.1:12345
```
Then get the log by:
```
soc/xtensa/intel_adsp/cavstool_client.py -s 192.168.0.1:54321
```

Or doing following command to output log immediately:
```
west flash --remote-host=192.168.0.1:12345 --pty=192.168.0.1:54321
```

Also add specifying the port and network interface for cavs-server:
```
cavstool.py -s 192.168.0.1:54321 --r 12345
OR
cavstool.py -s 192.168.0.1 -p 54321 --r 12345
```

Fixes #46865
Fixes #48584 

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

PS: this PR is spilt from https://github.com/zephyrproject-rtos/zephyr/pull/46880

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -115,7 +115,20 @@ def main():
+        PORT_REQ = int(h[1])
+
+    elif l == 2 or l > 3:
+        log.error("Please use format --remote-host=HOST or HOST:REQ_PORT:LOG_PORT")
```
Thanks for separating this from a previous, bigger and apparently not related PR. Copying my comment from the other PR:

> Please use a new --argument for this. Don't extend standard conventions; they're painful enough to parse already. The number of colons : is used as a trick to identify bracket-less IPv6 addresses which means {hostname}:{req_port}:{log_port} could be wrongly identified as an IPv6 address.

https://bugzilla.mindrot.org/show_bug.cgi?id=1602
https://chromium-review.googlesource.com/c/chromiumos/third_party/autotest/+/312719/3/server/base_utils.py

enjiamai replied:
Fixed it.

enjiamai commented:
Hi @marc-hb , I just re-implement it, according to your suggestion in the https://github.com/zephyrproject-rtos/zephyr/pull/46880. Use the --arguments to do it. Could you please re-review these two PRs, thank you very much!

marc-hb commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -61,6 +61,17 @@ also specify the rimage tool, config and key by:
+
+.. code-block:: console
+
+   sudo ./cavstool.py -s 192.168.0.2
```
Why no support for `(INADDR_ANY, port)`? This would be so much simpler. Binding can be a nice security feature in complex network configurations but most DUTs will have a single interface which means zero security value.


https://docs.python.org/3/library/socket.html#socket-families

enjiamai replied:
Good suggestion, let me take a look, thanks!

enjiamai replied:
Hi @marc-hb , I would like to put this in the next update, (include what you suggested using the high level-api, might need some time, thanks!)

marc-hb reviewed (dismissed):
Very clear documentation, thank you.

enjiamai commented:
Rebase and add version information. Now it is ready for SSH port support.

chen-png commented:
would better provide same port options in server side.
because the server is still listening on fixed port 9999 and 10000, if the port was changed to other ports on client side,  then users need to do a port forwarding to make it works.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -21,6 +21,7 @@
 start_output = False
 lock = threading.Lock()
 
+VERSION_INFO = "VER 1.0 (ABI V1)"
```
What ABI is this?

enjiamai replied:
Good question. You are right. This is a not proper name use. It's not binary interface. Will remove it. Just use version to identify is enough.

My original thinking is to tell the user that the client and server should use the same version. Ex.
Assume the V1 support:    --port-req {port}
And the V2 might support it in a different way. (INADDR_ANY, port),

marc-hb replied:
The client, server and all other scripts all live in the same repo so I don't see any compatibility problem: people MUST use scripts all from the same commit, end of story. Don't solve problems that you don't have.

marc-hb commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -159,3 +170,68 @@ field. For example:
+*******************
+
+In some case, we might need to access the service of the CAVS server by
+specifying a port. There are two service working in the server:
```
Please make it clearer that the default values will work just fine for most people and that they can skip reading this very long section and extra overhead.

enjiamai replied:
Yes, updated it.

marc-hb commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -61,6 +61,17 @@ also specify the rimage tool, config and key by:
               --key [path to signing key]
 
+
+Note: Sometimes you might have multiple network interfaces. In this case,
```
```suggestion
Note: the current server script has a design issue where it listens only to a random network interface. This can be a problem if you have multiple interfaces or want to use localhost, in this case....
```

marc-hb commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -61,6 +61,17 @@ also specify the rimage tool, config and key by:
+
+   sudo ./cavstool.py -s 192.168.0.2
+
+
```
In the future, the script will be fixed and will listen on all interfaces by default and this option will allow you to bind to a specific interface if you need to.

enjiamai replied:
very good suggestion, working on it.

enjiamai replied:
Added specifying a log port and req port for the cavs server:
```
cavstool.py -s {host}:{log port} --req-port {req port} 
OR 
cavstool.py -s {host} --log-port {log port} --req-port {req port}
```

enjiamai commented:
Thanks for the review! @marc-hb @chen-png , working on updating it. as you suggested, I think the way of specifying the port such as (INADDR_ANY, port). I should do it this time and not to change it again later.

marc-hb commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -124,6 +134,15 @@ def flash(self, **kwargs):
         self.run_cmd = ([f'{self.cavstool}','-s', f'{self.remote_host}', f'{send_bin_fw}'])
         self.log_cmd = ([f'{self.cavstool}','-s', f'{self.remote_host}', '-l'])
 
+        # Specify the ports of the req and log service if needed.
+        if self.port_req:
+            self.run_cmd.append('--port-req')
+            self.run_cmd.append(f'{self.port_req}')
+
+        if self.port_log:
+            self.log_cmd.append('--port-log')
+            self.log_cmd.append(f'{self.port_log}')
+
```
> Adding one leads to a never-ending treadmill of adding two, three, four, [...] .

Agree 100%

> like the way `--tool-opt`

I haven't look at this one in particular but the problem with these is always the same: whitespace and quoting. Example:

```
# is --opt2 for west or for subtool?
west  --subtool-opt --opt1 --opt2

# OK now it is for west. But then...
west  --subtool-opt '--opt1 --opt2'

# Now let's try this
west --subtool-opt '--opt1="-DFU1=bar1 -DFU2=bar2" --opt2'

subtool ERROR: FU1 cannot be 'bar1 -DFU2=bar2'
```
etc.

The only reliable solution on the CLI is `--` but it allows only one subtool.

There is only so much a CLI can "scale". Past that point configuration files are required.

mbolivar-nordic replied:
> I haven't look at this one in particular but the problem with these is always the same: whitespace and quoting.

I think you will find that the implelmentations I pointed to solve both of these problems:

```py
parser.add_argument('--tool-opt', default=[], action='append', ...)
```

You can basically give as many of these as you want, and the result will be a python list which can contain arbitrary characters, including whitespace characters.

marc-hb replied:
Thanks @mbolivar-nordic ! So just to make absolutely sure there is zero misunderstanding:
```
west ... --tool-opt '-DKEYWORD1=value with some white space' --tool-opt --opt2
```
Correct? A thumbs up should be enough :-)

marc-hb replied:
> west ... --tool-opt '-DKEYWORD1=value with some white space' --tool-opt --opt2

Of course this does not work because... "thank you `argparse`"? Example:

```
west update --fetch-opt --depth=5
west update: error: argument -o/--fetch-opt: expected one argument
```
Workaround:
```
west update --fetch-opt=--depth=5
```

Infamous https://bugs.python.org/issue9334 "argparse does not accept options taking arguments beginning with dash (regression from optparse)"

enjiamai replied:
Thank very much for suggestions from @marc-hb @mbolivar-nordic , got learned again. I updated it according your recommendation.

enjiamai commented:
Thanks very much @marc-hb, @mbolivar-nordic ! Actually I thought again and again and tried to find better way for supporting the port specifying, now I have two kinds of implementation for it:

Assume remote service host is 192.168.0.1,  req-port=12345 log-port=54321

# Specify the port Propose 1
Run by west:
```
west flash --remote-host=192.168.0.1 --tool-opt="-r 12345 -p 54321"
```
OR
```
west flash --remote-host=192.168.0.1 --tool-opt="--port-req 12345 --port-log 54321"
```

Run by twister:
```
twister -W --device-testing -p intel_adsp_cavs18 \
--west-flash="--remote-host=192.168.0.1,--tool-opt=-r 12345" \
--device-serial-pty="/home/user/zephyrproject/zephyr/soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.0.1,-r,54321,-l" \
-vv -T tests/kernel/semaphore/semaphore/
```

# Specify the port Propose 2
Run by west:
```
west flash --remote-host=192.168.0.1,12345
soc/xtensa/intel_adsp/cavstool_client.py -s 192.168.0.1,54321
```
OR
```
west flash --remote-host=192.168.0.1,12345 --pty=192.168.0.1,54321
```

Run by twister:
```
twister -W --device-testing -p intel_adsp_cavs18 \
--west-flash="--remote-host=192.168.0.1,12345" \
--device-serial-pty="/home/user/zephyrproject/zephyr/soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.0.1,-p,12345,-l" \
-vv -T tests/kernel/semaphore/semaphore/
```

Any suggestion for these ways of specifying the port? @marc-hb 
(I also added --tool-opt for both to extend the parameter for the future)

marc-hb commented:
> west flash --remote-host=192.168.0.1,12345 --pty=192.168.0.1,54321

The super-standard notation for port numbers is `:`, please stick to that as much as possible: `west flash --remote-host=192.168.0.1:12345 --pty=192.168.0.1:54321`

> west flash --remote-host=192.168.0.1 --tool-opt="-r 12345 -p 54321"

As discussed above, you cannot use whitespace separators inside arguments AND support whitespace in arguments. It leads to very confusing situations like this:

```
west flash --remote-host=192.168.0.1 --tool-opt="-r 12345 --arg='white space'"

=> tool_opt_args = [   "-r",    "12345",   "--arg='white",   "space'" ] # BAD
```

The solution shown by @mbolivar-nordic is verbose but it always works:

```
west flash --remote-host=192.168.0.1 --tool-opt=-r --tool-opt=12345 --tool-opt=--arg='white space'

=> tool_opt_args = [   "-r",    "12345",   "--arg=white space" ] # GOOD
```

> Another thing I want to highlight is that you can see I cannot keep the parameters consistent

Not sure what parsing twister does, I recommend getting it "right" in `west` first and then worry about twister and consistency later.

marc-hb commented:
> twister -W --device-testing -p intel_adsp_cavs18 \
--west-flash="--remote-host=192.168.0.1,12345" \
--device-serial-pty="/home/user/zephyrproject/zephyr/soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.0.1,-p,12345,-l" \
-vv -T tests/kernel/semaphore/semaphore/

The length and complexity of such a command line is crying for a configuration file.

enjiamai commented:
> The super-standard notation for port numbers is :, please stick to that as much as possible: west flash --remote-host=192.168.0.1:12345 --pty=192.168.0.1:54321

This is very important information for me! I got stocked for couple days and feel headache for what to use. If  {host}:{port} is used, all the problem I meet can be solved well. I will use this way.  Thank you! @marc-hb .

enjiamai committed:
doc: boards: update the usage of intel_adsp boards

Update the usage of intel_adsp boards according to adding
the port support.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

marc-hb commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -48,6 +50,10 @@ def __init__(self,
 
+        self.tool_opt = []
+        for opts in [shlex.split(opt) for opt in tool_opt]:
+            self.tool_opt += opts
```
This looks wrong to me, I suspect it leads to the situation already mentioned:
```
west flash --remote-host=192.168.0.1 --tool-opt="-r 12345 --arg='white space'"

=> tool_opt_args = [   "-r",    "12345",   "--arg='white",   "space'" ] # BAD
```
Did you try this?

I think you should remove this code and ask the users to do this instead:
```
west flash --remote-host=192.168.0.1 --tool-opt=-r --tool-opt=12345 --tool-opt=--arg='white space'

=> tool_opt_args = [   "-r",    "12345",   "--arg=white space" ] # GOOD
```
It is important to do this right considering the multiple levels of nesting: twister->west->cavstool

enjiamai replied:
Hi @marc-hb , I think I fixed according to the suggestion from you and @mbolivar-nordic , could you please take a look again? thanks!

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -133,9 +136,24 @@ def main():
-HOST = args.server_addr
+if args.server_addr:
+    try:
+        HOST, PORT = args.server_addr.split(":", 2)
```
This is not compatible with IPv6. Use `urlparse`:
```python
>>> from urllib.parse import urlparse

>>> urlparse('//[::1]:123').hostname
'::1'
>>> urlparse('//[::1]:123').port
123
>>> print(urlparse('//[::1]').port)
None
```

enjiamai replied:
This is really good for me, thanks for letting me know this!

enjiamai committed:
soc: xtensa: add support for specifying the service port

Add specifying the port for the remote cavstool remote service.
Ex.

    west flash --remote-host {host}:{port} \
               --pty {host}:{port}

Specify the port is optional when running west.

And another major change is now we have to specify the remote log
service by --pty to make it output the log message immediately.
Previously it will output directly. Why we make this change is
because this is more close to the behavior we use west flash.

Fixes. #46865

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
soc: xtensa: cavstool server support specifying the port and iface

Make the cavstool service can start with specifying the port
optionally by:

   cavstool.py -s {host}:{log port} --req-port {req port}
OR
   cavstool.py -s {host} --log-port {log port} --req-port {req port}

And we can also specify the network interface instead:

   cavstool.py -i {network iface}

If the server address or the network interface is not specified,
it will use INADDR_ANY by default.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai commented:
> would better provide same port options in server side. because the server is still listening on fixed port 9999 and 10000, if the port was changed to other ports on client side, then users need to do a port forwarding to make it works.

Hi @chen-png , I implemented the -i paramater for specifying the network interface and we can specify the log port and req port in both client and server side, could you please help review again? Thanks!

stephanosio commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -46,7 +67,18 @@ Build and run the tests
-   west flash --remote-host [remote hostname or ip addr]
+   west flash --remote-host {host}:{port} \
+              --pty {host}:{port}\
+Ex.
```
https://github.com/zephyrproject-rtos/zephyr/issues/48215#issuecomment-1193097630

enjiamai replied:
Thanks for your help. Got it fixed.

enjiamai commented:
I don't know why the re-request review button doesn't work for me. Excuse me @mbolivar-nordic @marc-hb @chen-png @stephanosio , could you please help re-review again, thank you very much!

marc-hb commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -159,3 +244,27 @@ field. For example:
+.. code-block:: console
+
+   west flash --remote-host=192.168.0.1 --pty=192.168.0.1 \
+              --log-opt=--arg='white space' --log-opt='-r 12345'
```
Would this make more sense?

```suggestion
              --log-opt=--arg='white space' --log-opt=-r --log-opt=12345
```

enjiamai replied:
Yes, I think so too. Got it updated.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -872,12 +882,23 @@ def get_host_ip():
 if args.server_addr:
-    HOST = args.server_addr
+    url = urlparse("//" + args.server_addr)
+    HOST, _, PORT = url.netloc.partition(':')
```
No, you must use `.hostname` and `.port`, you must not use `.netloc`. `urlparse` does all the parsing for you, do not second-guess `urlparse`

```python
# BAD
>>> urllib.parse.urlparse('//[::1]').netloc
'[::1]'
>>> urllib.parse.urlparse('//[::1]').netloc.partition(':')
('[', ':', ':1]') # BAD!
```

```python
>>> urllib.parse.urlparse('//[::1]').hostname
'::1'
>>> urllib.parse.urlparse('//[::1]').port
None
```

enjiamai replied:
It is much better, got it fixed, thanks.

marc-hb commented on boards/xtensa/intel_adsp_cavs15/doc/index.rst:
```diff
@@ -126,7 +126,7 @@ a path to the same key file used above.
     ./scripts/twister --device-testing -p intel_adsp_cavs15 \
       --device-serial-pty $ZEPHYR_BASE/soc/xtensa/intel_adsp/tools/cavstool_client.py,cavs15,-l \
-      --west-flash "--remote-host=cavs15,--pty"
+      --west-flash "--remote-host=cavs15"
```
```suggestion
      --west-flash "--remote-host=myboard.local"
```
Please replace `cavs15` as a sample hostname, it looks exactly like a board name and SoC series name.

marc-hb commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -68,9 +73,15 @@ def do_add_parser(cls, parser):
-                            help='the log will not output immediately to STDOUT, you \
-                            can redirect it to a serial PTY')
+        parser.add_argument('--pty',
+                            help='the log will be output immediately to STDOUT, the \
```
```suggestion
                            help='Capture the output of cavstool.py running on --remote-host and stream it remotely to west's standard output
```

BTW cavstool.py uses both stdout and stderr to print two different things. How are they forwarded here?

enjiamai replied:
That's indeed a problem now. Currently the log from stderr output to log at server-side. How do you think that forwarding them to log service also?

mbolivar-nordic replied:
@enjiamai I'm not sure if this comment is still open or not, but I'm going to leave it up to you to decide if this current implementation is OK as I don't know anything about cavstool.py

marc-hb commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -68,9 +73,15 @@ def do_add_parser(cls, parser):
+                            help='''Additional options for run/request service tool,
+                            e.g. \'--lock\' ''')
+        parser.add_argument('--log-opt', default=[], action='append',
+                            help='''Additional options for log service tool,
```
```suggestion
                            help='''Additional options for `cavstool`
```
Or is it for `cavstool_server`?

marc-hb commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -122,15 +135,25 @@ def flash(self, **kwargs):
         # not output to stdout directly. That means we can
         # make the log output to the PTY.
-        if not self.pty:
+        if self.pty is not None:
```
The comment above fails to make any difference between the target and build machine. this is very confusing.

enjiamai replied:
Thanks for pointing this out, I neglected updating this comment message. Got it fixed.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -856,7 +860,13 @@ def get_host_ip():
                 help="No current log buffer at start, just new output")
 ap.add_argument("-s", "--server-addr",
-                help="Specify the IP address that the server to active")
+                help="Specify the IP address that the log server to active")
```
```suggestion
                help="Specify the only IP address the log server will LISTEN on")
```

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -856,7 +860,13 @@ def get_host_ip():
-                help="Specify the IP address that the server to active")
+                help="Specify the IP address that the log server to active")
+ap.add_argument("-i", "--interface",
+                help="Specify the network interface where the service runs")
```
```suggestion
                help="Specify the network interface the service will LISTEN to")
```
A service never runs on a specific network interface. It can BIND to a single network interface to LISTEN to when it starts (not to random one by default, this is the current bug) and then the service completely forgets it did that.

enjiamai replied:
> A service never runs on a specific network interface. It can BIND to a single network interface to LISTEN to when it starts (not to random one by default, this is the current bug) and then the service completely forgets it did that.

Thank you! @marc-hb , how do you think if we make the user must select the network interface while there are more than one network interface existing (except localhost)?

marc-hb replied:
> Thank you! @marc-hb , how do you think if we make the user must select the network interface while there are more than one network interface existing (except localhost)?

I think we already discussed this. The fix is to bind to `INADDR_ANY` (= all interfaces) by default when nothing is provided on the command line. That's it.

marc-hb reviewed (dismissed):
What happened to this important documentation update I recommended earlier?

> Note: the current server script has a design issue where it listens only to a random network interface. This can be a problem if you have multiple interfaces or want to use localhost, in this case....

In the simplest cases where there is a single network interface this design issue is not a problem. Then the user should not feel like they have to provide the same IP address again and again, especially not considering the length and nesting levels of a typical command (`--log-opt` and what not)

I realize this issue and other documentation issues pre-date this PR but I don't think you can extend these features without fixing their documentation first. Otherwise the confusing documentation makes it really hard to review the feature changes.

mbolivar-nordic commented:
> The solution shown by @mbolivar-nordic is verbose but it always works

I would be open to reserving/enforcing `-O` as a short option for `--tool-opt` if you would find this useful. I've been waiting for someone to complain about the long option before giving it a short option.

enjiamai commented:
Appreciate again for reviewing it so quickly! @marc-hb, and also @mbolivar-nordic :)

enjiamai commented:
Hi @marc-hb, I tried to add the INADDR_ANY you mentioned previous, then did  some update in the documents. Could you please help review again when you get a change? Thank you!

enjiamai commented:
Hi @mbolivar-nordic , @stephanosio , could you please also help review this again when you get a change? Thank you!

enjiamai commented:
Excuse me, @mbolivar-nordic @marc-hb , could you please help to review this again, thank you very much!

mbolivar-nordic commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -69,8 +74,14 @@ def do_add_parser(cls, parser):
+        parser.add_argument('-o','--run-opt', default=[], action='append',
+                            help='''Additional options for run/request service tool,
+                            e.g. \'--lock\' ''')
+        parser.add_argument('-O', '--log-opt', default=[], action='append',
+                            help='''Additional options for `cavstool`,
+                            e.g. \'--timeout 60\' ''')
```
I am confused about this, @enjiamai . You gave a +1 to reserving `-O` for `--tool-opt` here:

https://github.com/zephyrproject-rtos/zephyr/pull/46972#issuecomment-1194509815

But here in the PR you are using `-O` for something different. We can't have both. What's going on?

mbolivar-nordic replied:
Here is a new pull request where we reserve and standardize on -O for --tool-opt, as previously agreed upon in this PR, for reference:

https://github.com/zephyrproject-rtos/zephyr/pull/48581

enjiamai replied:
Thanks! @mbolivar-nordic , working on updating and rebasing this PR, according to your feature extension in the west framework, :)

enjiamai committed:
soc: xtensa: tools: remove dependency of netifaces package

Because we use INADDR_ANY by default now, specifying a net interface
is no longer needed. Remove it and remove the dependency of the
netifaces python package.

Fixes. #48584

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
runner: intel_adsp: add parameters --tool-opt

Add parameter for possible extending in the future, --tool-opt
passes arguments to execution of the run service.
Ex.
  --tool-opt=--arg='white space' --tool-opt=-r --tool-opt=12345

will be parsed as ['--arg=white space', '-r', '12345']

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
runner: intel_adsp: no need to specify host in --pty if no port assign

No need to specify host again in --pty when not specifying the port,
slightly reduce the input.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai commented:
Update this PR focus on:
1. Address the comments from @mbolivar-nordic, adapting to https://github.com/zephyrproject-rtos/zephyr/pull/48581. To use --tool-opt.
2. Remove the dependency of netifaces package, to fix https://github.com/zephyrproject-rtos/zephyr/issues/48584
3. Update the doc accordingly.
4. Rebase to latest code.

Could you please help on reviewing it again? @mbolivar-nordic , @marc-hb , thank you!

marc-hb commented:
Very busy week sorry, will review the next force-push

enjiamai commented:
Excuse me, @mbolivar-nordic , could you please help on reviewing this again? Actually this is blocked some of our internal progress. Thank you so much!

laurenmurphyx64 commented on boards/xtensa/intel_adsp_cavs18/doc/generic_intel_adsp.rst:
```diff
@@ -27,6 +27,27 @@ Set up the environment
      https://github.com/thesofproject/rimage.
 
 
+How Remote Service works
```
Maybe I'm missing something, why do we have two separate docs, one for 15 and one for generic CAVS? Can we delete the 15 one or point it to the generic doc in 18?

enjiamai replied:
Thanks the suggestions! @laurenmurphyx64. I have another PR depending on this, so I will try to regular them after this PR.

laurenmurphyx64 commented:
@enjiamai I'm sorry this took so long for me to review. As for the HW pool, from experimenting on APL...

West works!
Twister works!

Really amazing work!!! :rocket: I did a couple pretty mean things - for example, I CTRL+C killed the client in the middle of running a test in a long Twister run, then rehit with West / restarted the Twister run and the server just got right back into it even with a BrokenPipeError on the console. Very impressive!

The only issue is the hardware map. Maybe this is just user error, in which case I'm ready to approve, but when I try to adapt your hardware map example for our hardware pool, I get this:

```
DEBUG   - Flash command: ['west', 'flash', '--skip-rebuild', '-d', '/home/herbert/zephyr-external/zephyr/twister-out/intel_adsp_cavs15/samples/hello_world/sample.basic.helloworld', '--runner', 'intel_adsp', '--remote-host=<USER>@<HOSTNAME>', '--port-req=<REQ_PORT>', '--port-log=<LOG_PORT>']
DEBUG   - DEVICE: usage: cavstool_client.py
DEBUG   - DEVICE: [-h]
DEBUG   - DEVICE: [-q]
DEBUG   - DEVICE: [-l]
DEBUG   - DEVICE: [-s SERVER_ADDR]
DEBUG   - DEVICE: [fw_file]
DEBUG   - DEVICE: cavstool_client.py: error: unrecognized arguments: --port-log
DEBUG   - -- west flash: using runner intel_adsp
```

This is my cavs.map file:
```
   - connected: true
     id: None
     platform: intel_adsp_cavs15
     product: None
     runner: intel_adsp
     serial_pty: "/home/herbert/zephyr-external/zephyr/soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,<USER>@<HOSTNAME>,--port-log,<LOG_PORT>,-l"
     runner_params:
     - --remote-host=zephyr@zephyrtest-blue.jf.intel.com
     - --port-req=<REQ_PORT>
     - --port-log=<LOG_PORT>
```

Looking a little into it.

enjiamai committed:
soc: xtensa: tools: add specifying log and request port for cavs client

Besides in a form of {host}:{port} to specifying the port using, we
also support specifying --log-port and --req-port parameters for the
cavstool client.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai commented:
>      runner_params:
>      - --remote-host=zephyr@zephyrtest-blue.jf.intel.com
>      - --port-req=<REQ_PORT>
>      - --port-log=<LOG_PORT>

Yes, this is incorrect and I should also update this after --tool-opt has been added. I got them fixed, thank you!
